src/controller.js

```diff
@@ -0,0 +1,601 @@
+      // The entire file must be lower than 5 TB
+      throw new Error('Object must total fewer than 5 TB'); 
+    }
+    let unsignedRequest = {
+      service: 's3',
+      region: this.region,
+      method: 'POST',
+      protocol: this.s3protocol,
+      hostname: this.__s3hostname(bucket),
+      path: `/${key}?uploads=`,
+      headers: {
+        'x-amz-meta-content-sha256': sha256,
+        'x-amz-meta-content-length': size,
+      }
+    };
+
+    // If we have permissions, set those values on the headers
+    if (permissions) {
+      let permHeaders = this.__determinePermissionsHeaders(permissions);
+      for (let tuple of permHeaders) {
+        unsignedRequest.headers[tuple[0]] = tuple[1];
+      }
+    }
+
+    let signedRequest = aws4.sign(unsignedRequest);
+
+    let response = await this.runner({
+      req: this.__serializeRequest(signedRequest)
+    });
+    
+    let parsedResponse = parseS3Response(response.body);
+
+    let uploadId = this.__getUploadId(parsedResponse, bucket, key);
+
+    if (response.statusCode === 200) {
+      return uploadId;
+    } else {
+      throw new Error('Could not initiate multipart upload');
+    }
+  }
+
+  /**
+   * Generate the general request data for uploading an individual part of a
+   * multipart upload.  The uploadId must be provided from a previous call to
+   * the S3 initate multipart API.  The parts argument is a list of objects
+   * describing the S3 upload parts.  They should be in the format {sha256:
+   * '...', offset: int, size: int}
+   *
+   * NOTE: It is required that the process that eventually runs this request
+   * parse the response to get the Etag for the purpose of completing the
+   * upload
+   *
+   * http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadUploadPart.html
+   */
+  async generateMultipartRequest(opts) {
+    opts = runSchema(opts, Joi.object().keys({
+      bucket: schemas.bucket.required(),
+      key: schemas.key.required(),
+      uploadId: Joi.string().required(),
+      parts: schemas.parts.required(),
+    }));
+
+    let {bucket, key, uploadId, parts} = opts;
+    let requests = [];
+
+    for (let num = 1 ; num <= parts.length ; num++) {
+      let part = parts[num - 1];
+
+      // This is sort of hard to encapsulate in Joi for me, so for the time
+      // being we'll leave this as an external check until I can figure out how
+      // to say that all parts until the last must be at least 5MB
+      if (part.size < 5 * 1024 * 1024 && num < parts.length) {
+        throw new Error(`Part ${num}/${parts.length} must be more than 5MB, except last`);
+      }
+
+      let signedRequest = aws4.sign({
+        service: 's3',
+        region: this.region,
+        method: 'PUT',
+        protocol: this.s3protocol,
+        hostname: this.__s3hostname(bucket),
+        path: `/${key}?partNumber=${num}&uploadId=${uploadId}`,
+        headers: {
+          'x-amz-content-sha256': part.sha256,
+          'content-length': part.size,
+        }
```

Comment:
that's sort of my thinking.  The only reason I wanted MD5s is to have the Content-MD5 header offered to clients, in the hope that because it's an HTTP spec and not a vendored prefix, it might trick clients into automatically doing MD5 sums for those who don't chose to do `x-amz-content-sha256` validation.